### Linear Regression Learning Practice

In [1]:
import pandas as pd
import numpy as np


### Import Data
Upload data, Set a path to Read Data



In [2]:
Path = r'C:\Users\PC\Downloads\archive\World-happiness-report-updated_2024.csv'
Data = pd.read_csv(Path, encoding='latin1')     #encoding= 'latin1' avoid uptf error

In [3]:
Data.head()

Country name  year  Life Ladder  Log GDP per capita  Social support  \
0  Afghanistan  2008        3.724               7.350           0.451   
1  Afghanistan  2009        4.402               7.509           0.552   
2  Afghanistan  2010        4.758               7.614           0.539   
3  Afghanistan  2011        3.832               7.581           0.521   
4  Afghanistan  2012        3.783               7.661           0.521   

   Healthy life expectancy at birth  Freedom to make life choices  Generosity  \
0                              50.5                         0.718       0.164   
1                              50.8                         0.679       0.187   
2                              51.1                         0.600       0.118   
3                              51.4                         0.496       0.160   
4                              51.7                         0.531       0.234   

   Perceptions of corruption  Positive affect  Negative affect  
0                      0.882            0.414            0.258  
1                      0.850            0.481            0.237  
2                      0.707            0.517            0.275  
3                      0.731            0.480            0.267  
4                      0.776            0.614            0.268

In [4]:
Data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2363 entries, 0 to 2362
Data columns (total 11 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Country name                      2363 non-null   object 
 1   year                              2363 non-null   int64  
 2   Life Ladder                       2363 non-null   float64
 3   Log GDP per capita                2335 non-null   float64
 4   Social support                    2350 non-null   float64
 5   Healthy life expectancy at birth  2300 non-null   float64
 6   Freedom to make life choices      2327 non-null   float64
 7   Generosity                        2282 non-null   float64
 8   Perceptions of corruption         2238 non-null   float64
 9   Positive affect                   2339 non-null   float64
 10  Negative affect                   2347 non-null   float64
dtypes: float64(9), int64(1), object(1)
memory usage: 203.2+ KB


In [5]:
Data.describe()

year  Life Ladder  Log GDP per capita  Social support  \
count  2363.000000  2363.000000         2335.000000     2350.000000   
mean   2014.763860     5.483566            9.399671        0.809369   
std       5.059436     1.125522            1.152069        0.121212   
min    2005.000000     1.281000            5.527000        0.228000   
25%    2011.000000     4.647000            8.506500        0.744000   
50%    2015.000000     5.449000            9.503000        0.834500   
75%    2019.000000     6.323500           10.392500        0.904000   
max    2023.000000     8.019000           11.676000        0.987000   

       Healthy life expectancy at birth  Freedom to make life choices  \
count                       2300.000000                   2327.000000   
mean                          63.401828                      0.750282   
std                            6.842644                      0.139357   
min                            6.720000                      0.228000   
25%                           59.195000                      0.661000   
50%                           65.100000                      0.771000   
75%                           68.552500                      0.862000   
max                           74.600000                      0.985000   

        Generosity  Perceptions of corruption  Positive affect  \
count  2282.000000                2238.000000      2339.000000   
mean      0.000098                   0.743971         0.651882   
std       0.161388                   0.184865         0.106240   
min      -0.340000                   0.035000         0.179000   
25%      -0.112000                   0.687000         0.572000   
50%      -0.022000                   0.798500         0.663000   
75%       0.093750                   0.867750         0.737000   
max       0.700000                   0.983000         0.884000   

       Negative affect  
count      2347.000000  
mean          0.273151  
std           0.087131  
min           0.083000  
25%           0.209000  
50%           0.262000  
75%           0.326000  
max           0.705000

### Pre Processing
Finding Null,empty etc and replacing it
Checking Data Types

In [6]:
Data["Life Ladder"].dtype

dtype('float64')

In [7]:
Data["Log GDP per capita"].dtype

dtype('float64')

In [8]:
Data["Life Ladder"].isnull().sum()

np.int64(0)

In [9]:
Data["Log GDP per capita"].isnull().sum()

np.int64(28)

In [10]:
Data.dropna(subset=["Log GDP per capita"], axis=0, inplace = True)    #T capital
Data["Log GDP per capita"].isnull().sum()

Data.dropna(subset=["Log GDP per capita"], axis=0, inplace=True)
Data["Log GDP per capita"].isnull().sum()            

np.int64(0)

### Using Scikit learn Library for Linear Regression


In [47]:
#import linear model from Scikit_learn
from sklearn.linear_model import LinearRegression   
#LinearRegression    no-
from sklearn.model_selection import train_test_split

In [48]:
#Create linear regression object
Linear_model = LinearRegression()

In [49]:
#define X as feature set, Y as target variabe
X= Data[["Log GDP per capita"]]     # X should be dataframe 2D , double square bracket make it
Y= Data["Life Ladder"]

In [50]:
Linear_model.fit(X,Y)

LinearRegression()

In [15]:
Linear_model.predict(X)

array([3.93151437, 4.05249435, 4.1323868 , ..., 4.16510465, 4.1749961 ,
       4.18184402], shape=(2335,))

In [16]:
print("Bo =", Linear_model.intercept_)         # Base   _
print("B1 =", Linear_model.coef_)           #m  Slope    _

Bo = -1.660956653772244
B1 = [0.76088041]


In [17]:
# Mean Squared Error (MSE)
Y_hat = Linear_model.predict(X)
mean_squared_error(Data["Life Ladder"], Y_hat)

NameError: name 'mean_squared_error' is not defined

$ Predicted\_Value = c_0 + c_1 \cdot Log\_GDP per capita $ <br>
$c_0 =-1.660956653772244$  <br>
$c_1 = 0.76088041$ <br>

$ Predicted\_Value = -1.660956653772244 + 0.76088041 \cdot Log\_GDP per capita $

In [ ]:
Linear_model.predict( np.array( [[7.350]]))   #predict background x-train,y-train
                                                #double bracket,  Input feature   m*x +b

In [ ]:
Linear_model.predict( np.array( [[7.6]]))

In [ ]:
Linear_model.predict(X)

Data["Social support"].dtype


In [18]:
Data["Social support"].isnull().sum()

np.int64(13)

In [19]:
Data.dropna(subset= ["Social support"] ,axis=0, inplace= True)

In [20]:
print(Data["Social support"].isnull().sum())

0


In [21]:
print(Data["Generosity"].isnull().sum())

48


In [22]:
Data.dropna(subset=["Generosity"], axis=0,inplace=True)
print(Data["Generosity"].isnull().sum())

0


In [23]:
print(Data["Freedom to make life choices"].isnull().sum())

25


In [24]:
Data.dropna(subset=["Freedom to make life choices"], axis=0, inplace= True)
print(Data["Freedom to make life choices"].isnull().sum())

0


In [25]:
Z= Data[["Social support", "Generosity", "Freedom to make life choices"]]

In [26]:
Y= Data["Life Ladder"]

In [27]:
Linear_model.fit(Z,Y)

LinearRegression()

In [28]:
print("B0_=",Linear_model.intercept_)          #intercept_
print("B_{1-3}",Linear_model.coef_)               #coef_

B0_= -0.7261430430457976
B_{1-3} [5.64450203 0.34411263 2.18409781]


In [29]:
Linear_model.predict(np.array([[0.552,0.187,0.679]]))

C:\Users\PC\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\LocalCache\local-packages\Python313\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([3.93697356])

In [30]:
Linear_model.predict(Z)

array([3.44414407, 3.93697356, 3.66730753, ..., 4.57213329, 4.43200686,
       4.77270949], shape=(2249,))

# Price per apple * No. of apple + Bag Cost  (analogy)

## How the value is predicted ##
## b0 + b1(social support) + b2(generosity) + b3( freedom to make choices
## -0.726 + 5.64(0.553) + 0.344(0.187) + 2.18(0.679)
 # 3.937 #

 # difference from actual value : 4.402- 3.937 = 0.46

 # Note : If slope decrease then b0- b1x


In [31]:
Data.head()

Country name  year  Life Ladder  Log GDP per capita  Social support  \
0  Afghanistan  2008        3.724               7.350           0.451   
1  Afghanistan  2009        4.402               7.509           0.552   
2  Afghanistan  2010        4.758               7.614           0.539   
3  Afghanistan  2011        3.832               7.581           0.521   
4  Afghanistan  2012        3.783               7.661           0.521   

   Healthy life expectancy at birth  Freedom to make life choices  Generosity  \
0                              50.5                         0.718       0.164   
1                              50.8                         0.679       0.187   
2                              51.1                         0.600       0.118   
3                              51.4                         0.496       0.160   
4                              51.7                         0.531       0.234   

   Perceptions of corruption  Positive affect  Negative affect  
0                      0.882            0.414            0.258  
1                      0.850            0.481            0.237  
2                      0.707            0.517            0.275  
3                      0.731            0.480            0.267  
4                      0.776            0.614            0.268

In [32]:
Linear_model.score(Z,Y)

0.5893336056706773

## Model Evaluation ##

# mse without train test

In [33]:
from sklearn.metrics import mean_squared_error

In [34]:
# Mean Squared Error (MSE)                #without traintest split
Y_hat = Linear_model.predict(Z)                 
mean_squared_error(Data["Life Ladder"], Y_hat)

0.515182199681637

# mse with train test split

In [39]:
from sklearn.model_selection import train_test_split

In [40]:
Z_train,Z_test,Y_train,Y_test= train_test_split(Z,Y, test_size=0.3, random_state=40)

In [41]:
Linear_model.fit(Z_train,Y_train)

LinearRegression()

In [42]:
Y_predict_test=Linear_model.predict(Z_test)    #z -- input

In [43]:
mse_test=mean_squared_error (Y_test, Y_predict_test)     #  Y --- predict (outyput)    ,y test origianl; y predict test predicted

In [44]:
print(mse_test)

0.5159756852082104


# R2 value (Perc. of variation in Y)
# higher r2 mean greater fit and best but note it for overfitting
# R2=   (1 - mse of reg. line/mse of avg of data)
# it used to test two model on same data to see which one is good
## R2 has values between 0 and 1

In [ ]:
from sklearn.metrics import r2_score

In [ ]:
Linear_model.score(Z,Y)     


In [53]:
Linear_model.predict(X,Y)

TypeError: LinearModel.predict() takes 2 positional arguments but 3 were given